Port ControlPart related controls to the new IPC serialization format

gavin-apple opened the pull request:
#### a7330e04e7107c53fc208a5cf3b7a79e7f80a7b2
<pre>
Port ControlPart related controls to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=261819">https://bugs.webkit.org/show_bug.cgi?id=261819</a>
<a href="https://rdar.apple.com/115781032">rdar://115781032</a>

Reviewed by Aditya Keerthi and Said Abou-Hallawa.

This change ports ControlPart and it&apos;s related appearances to the new IPC
serialization format. In order to do this, we&apos;ve had to extract the
ControlPartAppearance out of the ControlPart class itself. ControlPart now
acts as a wrapper around the the ControlPartAppearance, the ControlFactory,
and the PlatformControl. It&apos;s the responsibility of the ControlPartAppearance
to perform appearance specific operations on behald of the ControlPart. This
includes instantiating the appearance specific PlatformControl.

This change also adds the all of the new ControlPartAppearance to the
.serialization.in files, allowing them to be serialized within a
ControlPart. This includes:
    - ApplePayButtonAppearance,
    - ButtonAppearance,
    - ColorWellAppearance,
    - DefaultButtonAppearance,
    - ImageControlsButtonAppearance,
    - InnerSpinButtonAppearance,
    - MenuListButtonAppearance,
    - MenuListAppearance,
    - MeterAppearance,
    - ProgressBarAppearance,
    - PushButtonAppearance,
    - SearchFieldCancelButtonAppearance,
    - SearchFieldAppearance,
    - SearchFieldResultsButtonAppearance,
    - SearchFieldResultsDecorationAppearance,
    - SliderThumbHorizontalAppearance,
    - SliderThumbVerticalAppearance,
    - SliderTrackHorizontalAppearance,
    - SliderTrackVerticalAppearance,
    - SquareButtonAppearance,
    - ListBoxAppearance,
    - TextAreaAppearance,
    - TextFieldAppearance,
    - CheckboxAppearance,
    - RadioAppearance

* Source/WebCore/platform/graphics/cocoa/controls/ApplePayButtonCocoa.h:
* Source/WebCore/platform/graphics/cocoa/controls/ApplePayButtonCocoa.mm:
(WebCore::ApplePayButtonCocoa::ApplePayButtonCocoa):
(WebCore::ApplePayButtonCocoa::draw):
* Source/WebCore/platform/graphics/cocoa/controls/ControlFactoryCocoa.h:
* Source/WebCore/platform/graphics/cocoa/controls/ControlFactoryCocoa.mm:
(WebCore::ControlFactoryCocoa::createPlatformApplePayButton):
* Source/WebCore/platform/graphics/controls/ApplePayButtonPart.cpp:
(WebCore::ApplePayButtonPart::ApplePayButtonPart):
(WebCore::ApplePayButtonPart::createPlatformControl):
(WebCore::ApplePayButtonPart::create): Deleted.
* Source/WebCore/platform/graphics/controls/ApplePayButtonPart.h:
* Source/WebCore/platform/graphics/controls/ButtonPart.h:
(WebCore::ButtonPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/ColorWellPart.h:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlPart.cpp:
(WebCore::ControlPart::create):
(WebCore::ControlPart::ControlPart):
(WebCore::ControlPart::createPlatformControl):
(WebCore::ControlPart::style const):
(WebCore::ControlPart::type const):
* Source/WebCore/platform/graphics/controls/ControlPart.h:
(WebCore::ControlPart::type const): Deleted.
* Source/WebCore/platform/graphics/controls/ImageControlsButtonPart.h:
* Source/WebCore/platform/graphics/controls/InnerSpinButtonPart.h:
* Source/WebCore/platform/graphics/controls/MenuListButtonPart.h:
* Source/WebCore/platform/graphics/controls/MenuListPart.h:
* Source/WebCore/platform/graphics/controls/MeterPart.cpp:
(WebCore::MeterPart::MeterPart):
(WebCore::MeterPart::createPlatformControl):
(WebCore::MeterPart::create): Deleted.
* Source/WebCore/platform/graphics/controls/MeterPart.h:
* Source/WebCore/platform/graphics/controls/PlatformControl.h:
(WebCore::PlatformControl::owningPart const):
* Source/WebCore/platform/graphics/controls/ProgressBarPart.cpp:
(WebCore::ProgressBarPart::ProgressBarPart):
(WebCore::ProgressBarPart::createPlatformControl):
(WebCore::ProgressBarPart::create): Deleted.
* Source/WebCore/platform/graphics/controls/ProgressBarPart.h:
(WebCore::ProgressBarPart::position const): Deleted.
(WebCore::ProgressBarPart::setPosition): Deleted.
(WebCore::ProgressBarPart::animationStartTime const): Deleted.
(WebCore::ProgressBarPart::setAnimationStartTime): Deleted.
* Source/WebCore/platform/graphics/controls/SearchFieldCancelButtonPart.h:
* Source/WebCore/platform/graphics/controls/SearchFieldPart.h:
* Source/WebCore/platform/graphics/controls/SearchFieldResultsPart.h:
(WebCore::SearchFieldResultsPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/SliderThumbPart.h:
(WebCore::SliderThumbPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/SliderTrackPart.cpp:
(WebCore::SliderTrackHorizontalPart::SliderTrackHorizontalPart):
(WebCore::SliderTrackVerticalPart::SliderTrackVerticalPart):
(WebCore::SliderTrackPart::SliderTrackPart):
(WebCore::SliderTrackHorizontalPart::drawTicks const):
(WebCore::SliderTrackVerticalPart::drawTicks const):
(WebCore::SliderTrackPart::createPlatformControl):
(WebCore::SliderTrackPart::create): Deleted.
(WebCore::SliderTrackPart::drawTicks const): Deleted.
* Source/WebCore/platform/graphics/controls/SliderTrackPart.h:
(isType): Deleted.
* Source/WebCore/platform/graphics/controls/TextAreaPart.h:
(WebCore::TextAreaPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/TextFieldPart.h:
* Source/WebCore/platform/graphics/controls/ToggleButtonPart.h:
(WebCore::ToggleButtonPart::createPlatformControl):
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.h:
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.mm:
(WebCore::ControlFactoryIOS::createPlatformButton):
(WebCore::ControlFactoryIOS::createPlatformColorWell):
(WebCore::ControlFactoryIOS::createPlatformInnerSpinButton):
(WebCore::ControlFactoryIOS::createPlatformMenuList):
(WebCore::ControlFactoryIOS::createPlatformMenuListButton):
(WebCore::ControlFactoryIOS::createPlatformMeter):
(WebCore::ControlFactoryIOS::createPlatformProgressBar):
(WebCore::ControlFactoryIOS::createPlatformSearchField):
(WebCore::ControlFactoryIOS::createPlatformSearchFieldCancelButton):
(WebCore::ControlFactoryIOS::createPlatformSearchFieldResults):
(WebCore::ControlFactoryIOS::createPlatformSliderThumb):
(WebCore::ControlFactoryIOS::createPlatformSliderTrack):
(WebCore::ControlFactoryIOS::createPlatformTextArea):
(WebCore::ControlFactoryIOS::createPlatformTextField):
(WebCore::ControlFactoryIOS::createPlatformToggleButton):
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm:
(WebCore::ButtonMac::ButtonMac):
* Source/WebCore/platform/graphics/mac/controls/ColorWellMac.h:
* Source/WebCore/platform/graphics/mac/controls/ColorWellMac.mm:
(WebCore::ColorWellMac::ColorWellMac):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::createPlatformButton):
(WebCore::ControlFactoryMac::createPlatformColorWell):
(WebCore::ControlFactoryMac::createPlatformImageControlsButton):
(WebCore::ControlFactoryMac::createPlatformInnerSpinButton):
(WebCore::ControlFactoryMac::createPlatformMenuList):
(WebCore::ControlFactoryMac::createPlatformMenuListButton):
(WebCore::ControlFactoryMac::createPlatformMeter):
(WebCore::ControlFactoryMac::createPlatformProgressBar):
(WebCore::ControlFactoryMac::createPlatformSearchField):
(WebCore::ControlFactoryMac::createPlatformSearchFieldCancelButton):
(WebCore::ControlFactoryMac::createPlatformSearchFieldResults):
(WebCore::ControlFactoryMac::createPlatformSliderThumb):
(WebCore::ControlFactoryMac::createPlatformSliderTrack):
(WebCore::ControlFactoryMac::createPlatformTextArea):
(WebCore::ControlFactoryMac::createPlatformTextField):
(WebCore::ControlFactoryMac::createPlatformToggleButton):
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm:
(WebCore::ImageControlsButtonMac::ImageControlsButtonMac):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
(WebCore::InnerSpinButtonMac::InnerSpinButtonMac):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::MenuListButtonMac::MenuListButtonMac):
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
(WebCore::MenuListMac::MenuListMac):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h:
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
(WebCore::MeterMac::MeterMac):
(WebCore::MeterMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.h:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
(WebCore::ProgressBarMac::ProgressBarMac):
(WebCore::ProgressBarMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm:
(WebCore::SearchFieldCancelButtonMac::SearchFieldCancelButtonMac):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.mm:
(WebCore::SearchFieldMac::SearchFieldMac):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.mm:
(WebCore::SearchFieldResultsMac::SearchFieldResultsMac):
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
(WebCore::SliderThumbMac::SliderThumbMac):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::SliderTrackMac::SliderTrackMac):
(WebCore::SliderTrackMac::rectForBounds const):
(WebCore::SliderTrackMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextAreaMac.h:
* Source/WebCore/platform/graphics/mac/controls/TextAreaMac.mm:
(WebCore::TextAreaMac::TextAreaMac):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm:
(WebCore::TextFieldMac::TextFieldMac):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm:
(WebCore::ToggleButtonMac::ToggleButtonMac):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::updateControlPartForRenderer const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/270445@main">https://commits.webkit.org/270445@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35b2b0fb05023aa9e2e8db6a65f5a49ae44a34f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26712 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27543 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23322 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25706 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1467 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23492 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25675 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2975 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21938 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28121 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2638 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22883 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28999 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23202 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23247 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26821 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2626 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/879 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3997 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6118 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3076 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2965 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/cocoa/controls/ApplePayButtonCocoa.mm:
```diff
@@ -105,16 +106,17 @@ static PKPaymentButtonStyle toPKPaymentButtonStyle(ApplePayButtonStyle style)
     auto logicalRect = borderRect.rect();
-    const auto& applePayButtonPart = owningApplePayButtonPart();
+
+    const auto& applePayButtonAppearance = std::get<ApplePayButtonAppearance>(owningPart().appearance());
```
How about this?

```
const auto& applePayButtonAppearance = owningPart().get<ApplePayButtonAppearance>();
```

This will be read like this: _From the owningPart get the ApplePayButtonAppearance_.

And the `ControlPart` method can be like this:

```
template <typename AppearanceType>
const AppearanceType& get() const
{
    ASSERT(AppearanceType::appearance == m_owningPart.type());
    return std::get<AppearanceType>(m_appearance);
}
```

shallawa commented on Source/WebCore/platform/graphics/controls/ApplePayButtonAppearance.cpp:
```diff
@@ -24,35 +24,31 @@
+ApplePayButtonAppearance::ApplePayButtonAppearance()
+    : m_buttonType(ApplePayButtonType::Plain)
+    , m_buttonStyle(ApplePayButtonStyle::White)
+    , m_locale()
```
You can call the other constructer instead:
```
ApplePayButtonAppearance::ApplePayButtonAppearance()
    : ApplePayButtonAppearance(ApplePayButtonType::Plain, ApplePayButtonStyle::White, { })
{
}
```

Or you initialize the members with these values 

```
    ApplePayButtonType m_buttonType { ApplePayButtonType::Plain };
    ApplePayButtonStyle m_buttonStyle { ApplePayButtonStyle::White };
```

And add the default constructor:

```
ApplePayButtonAppearance() = default;
```

shallawa commented on Source/WebCore/platform/graphics/controls/ApplePayButtonAppearance.h:
```diff
@@ -27,7 +27,9 @@
 
-#include "ControlPart.h"
+#include "ControlFactory.h"
+#include "PlatformControl.h"
```
I think you do not need to add these header files here. You can just add the forward declaration below:

```
class ControlFactory;
class PlatformControl;
```

shallawa commented on Source/WebCore/platform/graphics/controls/ButtonAppearance.h:
```diff
@@ -0,0 +1,69 @@
+class ControlFactory;
+class ControlPart;
+
+class ButtonPart {
```
ButtonPart -> ButtonAppearanceBase

shallawa commented on Source/WebCore/platform/graphics/controls/ButtonAppearance.h:
```diff
@@ -0,0 +1,69 @@
+    }
+};
+
+class ButtonButtonAppearance final : public ButtonPart {
```
ButtonButtonAppearance -> ButtonAppearance

shallawa commented on Source/WebCore/platform/graphics/controls/ButtonAppearance.h:
```diff
@@ -0,0 +1,69 @@
+public:
+    SquareButtonAppearance() = default;
+    static constexpr StyleAppearance appearance = StyleAppearance::SquareButton;
+};
```
Ideally this should be a template instead of one base class and four superclasses:
```
template<typename ButtonType>
class ButtonPart {
    ButtonAppearance()
    {
        COMPILE_ASSERT(ButtonType == StyleAppearance::Button || ...);
    }
    static constexpr StyleAppearance appearance = ButtonType;
};
```
But I am not sure whether CoreIPC supports class template encoding or not.

gavin-apple replied:
It has limited support for templated types. We don't currently have a way to template based on values, so I don't think ^ will work. Happy to look at adding it and refactoring later though.

shallawa commented on Source/WebCore/platform/graphics/controls/ControlPart.h:
```diff
@@ -36,11 +60,46 @@ class FloatRect;
 class GraphicsContext;
 class ControlFactory;
 
+using ControlAppearance = std::variant<
```
I think we should move this variant along with the appearance header files above to its own header file `ControlAppearance.h`

shallawa commented on Source/WebCore/platform/graphics/controls/ButtonAppearance.h:
```diff
@@ -0,0 +1,69 @@
+public:
+    std::unique_ptr<PlatformControl> createPlatformControl(ControlPart& part, ControlFactory& controlFactory)
+    {
+        return controlFactory.createPlatformButton(part);
```
I think we can omit the `ControlFactory` and use it directly from the `ControlPart`

```
    return part.controlFactory().createPlatformButton(part);
```

gavin-apple replied:
If we do this we create a cyclic dependency between ***Appearance.h -> ControlPart.h -> ControlAppearance.h -> ***Appearance.h. Today, many of the ***Appearance.h files don't have cxx counterparts. So it would mean we'd end up creating a bunch more files. Do you still want to do this?

shallawa commented on Source/WebCore/platform/graphics/controls/ControlFactory.h:
```diff
@@ -27,24 +27,9 @@
-class MenuListPart;
+class ControlPart;
+class PlatformControl;
 class PlatformControl;
```
Two forward declarations for `PlatformControl`

shallawa commented on Source/WebCore/platform/graphics/controls/ColorWellAppearance.h:
```diff
@@ -28,26 +28,22 @@
-        : ControlPart(StyleAppearance::ColorWell)
-    {
-    }
+class ColorWellAppearance final {
```
Why do we have to define this class as `final`?

shallawa commented on Source/WebCore/platform/graphics/controls/ControlAppearance.h:
```diff
@@ -0,0 +1,88 @@
+#include "SliderTrackAppearance.h"
+#include "TextAreaAppearance.h"
+#include "TextFieldAppearance.h"
+#include "ToggleButtonAppearance.h"
```
Can these replaced by the forward declaration of these classes?

gavin-apple replied:
If we do, then we're still going to need to have the full declaration available in ControlPart.h aren't we?

shallawa replied:
Yes you are right. Please ignore this comment.

shallawa commented on Source/WebCore/platform/graphics/controls/MeterAppearance.h:
```diff
@@ -25,20 +25,27 @@
 #pragma once
 
-#include "ControlPart.h"
+#include "ControlFactory.h"
```
Why do we `#include "ControlFactory.h"` here?

shallawa commented on Source/WebCore/platform/graphics/controls/ProgressBarAppearance.h:
```diff
@@ -25,15 +25,22 @@
 #pragma once
 
-#include "ControlPart.h"
+#include "ControlFactory.h"
```
Ditto. I think it can be forward declared only in this header file.

shallawa commented on Source/WebCore/platform/graphics/controls/SliderTrackAppearance.cpp:
```diff
@@ -0,0 +1,146 @@
+SliderTrackVerticalAppearance::SliderTrackVerticalAppearance(const IntSize& thumbSize, const IntRect& trackBounds, Vector<double>&& tickRatios)
+    : SliderTrackAppearance(thumbSize, trackBounds, WTFMove(tickRatios))
+{
+}
```
These constructors are not needed. Just add this line in the class `SliderTrackAppearance` as a protected member
```
SliderTrackAppearance() = default;
```

And add this line in the classes `SliderTrackHorizontalAppearance` and `SliderTrackVerticalAppearance` as a public member
```
using SliderTrackAppearance::SliderTrackAppearance;
```

gavin-apple replied:
Makes sense!

shallawa commented on Source/WebCore/platform/graphics/controls/SliderTrackAppearance.cpp:
```diff
@@ -0,0 +1,146 @@
+        context.fillRect(tickRect);
+    }
+}
+#endif
```
I do not know the constraints of the CoreIPC but this is too much code to be duplicated. Can we get part or parts of this code to be shared between these two classes?

shallawa replied:
Maybe something like this can work?
```
void SliderTrackAppearance::drawTicks(GraphicsContext& context, const FloatRect& rect, const ControlStyle& style, StyleAppearance appearance) const
{
    ...
    bool isHorizontal = appearance == StyleAppearance::SliderHorizontal;
    ...
}
```

```
void SliderTrackVerticalAppearance::drawTicks(GraphicsContext& context, const FloatRect& rect, const ControlStyle& style) const
{
    SliderTrackAppearance::drawTicks(context, rect, style, appearance);
}
```

```
void SliderTrackVerticalAppearance::drawTicks(GraphicsContext& context, const FloatRect& rect, const ControlStyle& style) const
{
    SliderTrackAppearance::drawTicks(context, rect, style, appearance);
}
```

shallawa commented on Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
```diff
@@ -76,10 +77,11 @@ static void trackGradientInterpolate(void*, const CGFloat* inData, CGFloat* outD
-    sliderTrackPart.drawTicks(context, borderRect.rect(), style);
+    WTF::switchOn(owningPart().appearance(),
+        [&](SliderTrackHorizontalAppearance& part) { return part.drawTicks(context, borderRect.rect(), style); },
+        [&](SliderTrackVerticalAppearance& part) { return part.drawTicks(context, borderRect.rect(), style); },
```
part -> appearance

shallawa replied:
Also using WTF::switchOn() seems no suitable here.  We know in advance there are only two possibilities for the appearance. How about this?
```
if (owningPart().type() == StyleAppearance::SliderHorizontal) {
    auto& appearance = owningPart().get<SliderTrackHorizontalAppearance>();
    appearance.drawTicks(context, borderRect.rect(), style);
} else {
    auto& appearance = owningPart().get<SliderTrackVerticalAppearance>();
    appearance.drawTicks(context, borderRect.rect(), style);
}
```

shallawa commented on Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
```diff
@@ -90,7 +92,7 @@ static void trackGradientInterpolate(void*, const CGFloat* inData, CGFloat* outD
     RetainPtr<CGFunctionRef> mainFunction = adoptCF(CGFunctionCreate(NULL, 1, NULL, 4, NULL, &mainCallbacks));
     RetainPtr<CGShadingRef> mainShading;
-    if (sliderTrackPart.type() == StyleAppearance::SliderVertical)
+    if (std::holds_alternative<SliderTrackVerticalAppearance>(owningPart().appearance()))
```
Can't this just 
```
if (owningPart().type() == StyleAppearance::SliderVertical)
```

shallawa commented on Source/WebCore/platform/graphics/controls/SliderTrackAppearance.h:
```diff
@@ -44,22 +54,37 @@ class SliderTrackPart : public ControlPart {
+class SliderTrackHorizontalAppearance final : public SliderTrackAppearance {
+public:
+    using SliderTrackAppearance::SliderTrackAppearance;
+    WEBCORE_EXPORT SliderTrackHorizontalAppearance(const IntSize& thumbSize, const IntRect& trackBounds, Vector<double>&& tickRatios);
```
I think this constructor can be deleted. The above `using SliderTrackAppearance::SliderTrackAppearance;` will allow using the two base class constructors.

shallawa commented on Source/WebCore/platform/graphics/controls/SliderTrackAppearance.h:
```diff
@@ -44,22 +54,37 @@ class SliderTrackPart : public ControlPart {
+class SliderTrackVerticalAppearance final : public SliderTrackAppearance {
+public:
+    using SliderTrackAppearance::SliderTrackAppearance;
+    WEBCORE_EXPORT SliderTrackVerticalAppearance(const IntSize& thumbSize, const IntRect& trackBounds, Vector<double>&& tickRatios);
```
Ditto.

shallawa commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -679,26 +687,35 @@ RefPtr<ControlPart> RenderTheme::createControlPart(const RenderObject& renderer)
+        ASSERT(part.type() == StyleAppearance::Meter && is<RenderMeter>(renderer));
+        updateMeterAppearanceForRenderer(std::get<MeterAppearance>(appearance), downcast<RenderMeter>(renderer));
         return;
     }
```
```
if (part.type() == StyleAppearance::Meter) {
    ASSERT(is<RenderMeter>(renderer));
    updateMeterAppearanceForRenderer(part.get<MeterAppearance>(), downcast<RenderMeter>(renderer));
    return;
}
```
`ControlPart::get<>()` have the first half of the assertion.

shallawa commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -679,26 +687,35 @@ RefPtr<ControlPart> RenderTheme::createControlPart(const RenderObject& renderer)
+        ASSERT(part.type() == StyleAppearance::ProgressBar && is<RenderProgress>(renderer));
+        updateProgressBarAppearanceForRenderer(std::get<ProgressBarAppearance>(appearance), downcast<RenderProgress>(renderer));
         return;
     }
```
Ditto.

shallawa commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -3161,37 +3116,155 @@ enum class WebCore::ApplePayButtonStyle : uint8_t {
+};
+#endif
+
+using WebCore::ControlAppearance = std::variant<
```
Why do we have to repeat this definition here? I was able to encode the DisplayListItem which is also variant without repeating its definition. See the changes in DisplayListArgumentCoders.serialization.in in https://github.com/WebKit/WebKit/pull/16688.

gavin-apple replied:
So we can expose it to the IPCTestingAPI

gavin-apple replied:
```
def generate_serialized_type_info(serialized_types, serialized_enums, headers, using_statements):
...
    for using_statement in using_statements:
        if using_statement.condition is not None:
            result.append('#if ' + using_statement.condition)
        result.append('        { "' + using_statement.name + '"_s, {')
        result.append('            { "' + using_statement.alias + '"_s, "alias"_s }')
        result.append('        } },')
        if using_statement.condition is not None:
            result.append('#endif')

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
```diff
@@ -33,21 +33,22 @@
     float trackWidth = sliderTrackWidth * style.zoomFactor;
 
-    auto& sliderTrackPart = owningSliderTrackPart();
+    auto& SliderTrackAppearance = owningPart();
```
```suggestion
    auto& sliderTrackAppearance = owningPart();
```

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
```diff
@@ -33,21 +33,22 @@
     
     // Set the height/width and align the location in the center of the difference.
-    if (sliderTrackPart.type() == StyleAppearance::SliderHorizontal) {
+    if (SliderTrackAppearance.type() == StyleAppearance::SliderHorizontal) {
```
```suggestion
    if (sliderTrackAppearance.type() == StyleAppearance::SliderHorizontal) {
```

pxlcoder commented on Source/WebCore/platform/graphics/controls/ToggleButtonAppearance.h:
```diff
@@ -0,0 +1,53 @@
+class RadioAppearance final : public ToggleButtonAppearance {
+public:
+    static constexpr StyleAppearance appearance = StyleAppearance::Checkbox;
+};
```
```suggestion
class RadioAppearance final : public ToggleButtonAppearance {
public:
    static constexpr StyleAppearance appearance = StyleAppearance::Radio;
};
```

pxlcoder reviewed (changes requested):
Requesting changes, as there appears to be an error with radio buttons.

shallawa commented on Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
```diff
@@ -33,21 +33,22 @@
     float trackWidth = sliderTrackWidth * style.zoomFactor;
 
-    auto& sliderTrackPart = owningSliderTrackPart();
+    auto& sliderTrackAppearance = owningPart();
```
Why did we change the name to `sliderTrackAppearance`? `owningPart()` returns a `ControlPart` not a `ControlAppearance`?

gavin-apple replied:
Fixed

gavin-apple committed:
Port ControlPart related controls to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=261819
rdar://115781032

Reviewed by Aditya Keerthi and Said Abou-Hallawa.

This change ports ControlPart and it's related appearances to the new IPC
serialization format. In order to do this, we've had to extract the
ControlPartAppearance out of the ControlPart class itself. ControlPart now
acts as a wrapper around the the ControlPartAppearance, the ControlFactory,
and the PlatformControl. It's the responsibility of the ControlPartAppearance
to perform appearance specific operations on behald of the ControlPart. This
includes instantiating the appearance specific PlatformControl.

This change also adds the all of the new ControlPartAppearance to the
.serialization.in files, allowing them to be serialized within a
ControlPart. This includes:
    - ApplePayButtonAppearance,
    - ButtonAppearance,
    - ColorWellAppearance,
    - DefaultButtonAppearance,
    - ImageControlsButtonAppearance,
    - InnerSpinButtonAppearance,
    - MenuListButtonAppearance,
    - MenuListAppearance,
    - MeterAppearance,
    - ProgressBarAppearance,
    - PushButtonAppearance,
    - SearchFieldCancelButtonAppearance,
    - SearchFieldAppearance,
    - SearchFieldResultsButtonAppearance,
    - SearchFieldResultsDecorationAppearance,
    - SliderThumbHorizontalAppearance,
    - SliderThumbVerticalAppearance,
    - SliderTrackHorizontalAppearance,
    - SliderTrackVerticalAppearance,
    - SquareButtonAppearance,
    - ListBoxAppearance,
    - TextAreaAppearance,
    - TextFieldAppearance,
    - CheckboxAppearance,
    - RadioAppearance

* Source/WebCore/platform/graphics/cocoa/controls/ApplePayButtonCocoa.h:
* Source/WebCore/platform/graphics/cocoa/controls/ApplePayButtonCocoa.mm:
(WebCore::ApplePayButtonCocoa::ApplePayButtonCocoa):
(WebCore::ApplePayButtonCocoa::draw):
* Source/WebCore/platform/graphics/cocoa/controls/ControlFactoryCocoa.h:
* Source/WebCore/platform/graphics/cocoa/controls/ControlFactoryCocoa.mm:
(WebCore::ControlFactoryCocoa::createPlatformApplePayButton):
* Source/WebCore/platform/graphics/controls/ApplePayButtonPart.cpp:
(WebCore::ApplePayButtonPart::ApplePayButtonPart):
(WebCore::ApplePayButtonPart::createPlatformControl):
(WebCore::ApplePayButtonPart::create): Deleted.
* Source/WebCore/platform/graphics/controls/ApplePayButtonPart.h:
* Source/WebCore/platform/graphics/controls/ButtonPart.h:
(WebCore::ButtonPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/ColorWellPart.h:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlPart.cpp:
(WebCore::ControlPart::create):
(WebCore::ControlPart::ControlPart):
(WebCore::ControlPart::createPlatformControl):
(WebCore::ControlPart::style const):
(WebCore::ControlPart::type const):
* Source/WebCore/platform/graphics/controls/ControlPart.h:
(WebCore::ControlPart::type const): Deleted.
* Source/WebCore/platform/graphics/controls/ImageControlsButtonPart.h:
* Source/WebCore/platform/graphics/controls/InnerSpinButtonPart.h:
* Source/WebCore/platform/graphics/controls/MenuListButtonPart.h:
* Source/WebCore/platform/graphics/controls/MenuListPart.h:
* Source/WebCore/platform/graphics/controls/MeterPart.cpp:
(WebCore::MeterPart::MeterPart):
(WebCore::MeterPart::createPlatformControl):
(WebCore::MeterPart::create): Deleted.
* Source/WebCore/platform/graphics/controls/MeterPart.h:
* Source/WebCore/platform/graphics/controls/PlatformControl.h:
(WebCore::PlatformControl::owningPart const):
* Source/WebCore/platform/graphics/controls/ProgressBarPart.cpp:
(WebCore::ProgressBarPart::ProgressBarPart):
(WebCore::ProgressBarPart::createPlatformControl):
(WebCore::ProgressBarPart::create): Deleted.
* Source/WebCore/platform/graphics/controls/ProgressBarPart.h:
(WebCore::ProgressBarPart::position const): Deleted.
(WebCore::ProgressBarPart::setPosition): Deleted.
(WebCore::ProgressBarPart::animationStartTime const): Deleted.
(WebCore::ProgressBarPart::setAnimationStartTime): Deleted.
* Source/WebCore/platform/graphics/controls/SearchFieldCancelButtonPart.h:
* Source/WebCore/platform/graphics/controls/SearchFieldPart.h:
* Source/WebCore/platform/graphics/controls/SearchFieldResultsPart.h:
(WebCore::SearchFieldResultsPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/SliderThumbPart.h:
(WebCore::SliderThumbPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/SliderTrackPart.cpp:
(WebCore::SliderTrackHorizontalPart::SliderTrackHorizontalPart):
(WebCore::SliderTrackVerticalPart::SliderTrackVerticalPart):
(WebCore::SliderTrackPart::SliderTrackPart):
(WebCore::SliderTrackHorizontalPart::drawTicks const):
(WebCore::SliderTrackVerticalPart::drawTicks const):
(WebCore::SliderTrackPart::createPlatformControl):
(WebCore::SliderTrackPart::create): Deleted.
(WebCore::SliderTrackPart::drawTicks const): Deleted.
* Source/WebCore/platform/graphics/controls/SliderTrackPart.h:
(isType): Deleted.
* Source/WebCore/platform/graphics/controls/TextAreaPart.h:
(WebCore::TextAreaPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/TextFieldPart.h:
* Source/WebCore/platform/graphics/controls/ToggleButtonPart.h:
(WebCore::ToggleButtonPart::createPlatformControl):
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.h:
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.mm:
(WebCore::ControlFactoryIOS::createPlatformButton):
(WebCore::ControlFactoryIOS::createPlatformColorWell):
(WebCore::ControlFactoryIOS::createPlatformInnerSpinButton):
(WebCore::ControlFactoryIOS::createPlatformMenuList):
(WebCore::ControlFactoryIOS::createPlatformMenuListButton):
(WebCore::ControlFactoryIOS::createPlatformMeter):
(WebCore::ControlFactoryIOS::createPlatformProgressBar):
(WebCore::ControlFactoryIOS::createPlatformSearchField):
(WebCore::ControlFactoryIOS::createPlatformSearchFieldCancelButton):
(WebCore::ControlFactoryIOS::createPlatformSearchFieldResults):
(WebCore::ControlFactoryIOS::createPlatformSliderThumb):
(WebCore::ControlFactoryIOS::createPlatformSliderTrack):
(WebCore::ControlFactoryIOS::createPlatformTextArea):
(WebCore::ControlFactoryIOS::createPlatformTextField):
(WebCore::ControlFactoryIOS::createPlatformToggleButton):
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm:
(WebCore::ButtonMac::ButtonMac):
* Source/WebCore/platform/graphics/mac/controls/ColorWellMac.h:
* Source/WebCore/platform/graphics/mac/controls/ColorWellMac.mm:
(WebCore::ColorWellMac::ColorWellMac):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::createPlatformButton):
(WebCore::ControlFactoryMac::createPlatformColorWell):
(WebCore::ControlFactoryMac::createPlatformImageControlsButton):
(WebCore::ControlFactoryMac::createPlatformInnerSpinButton):
(WebCore::ControlFactoryMac::createPlatformMenuList):
(WebCore::ControlFactoryMac::createPlatformMenuListButton):
(WebCore::ControlFactoryMac::createPlatformMeter):
(WebCore::ControlFactoryMac::createPlatformProgressBar):
(WebCore::ControlFactoryMac::createPlatformSearchField):
(WebCore::ControlFactoryMac::createPlatformSearchFieldCancelButton):
(WebCore::ControlFactoryMac::createPlatformSearchFieldResults):
(WebCore::ControlFactoryMac::createPlatformSliderThumb):
(WebCore::ControlFactoryMac::createPlatformSliderTrack):
(WebCore::ControlFactoryMac::createPlatformTextArea):
(WebCore::ControlFactoryMac::createPlatformTextField):
(WebCore::ControlFactoryMac::createPlatformToggleButton):
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm:
(WebCore::ImageControlsButtonMac::ImageControlsButtonMac):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
(WebCore::InnerSpinButtonMac::InnerSpinButtonMac):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::MenuListButtonMac::MenuListButtonMac):
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
(WebCore::MenuListMac::MenuListMac):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h:
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
(WebCore::MeterMac::MeterMac):
(WebCore::MeterMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.h:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
(WebCore::ProgressBarMac::ProgressBarMac):
(WebCore::ProgressBarMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm:
(WebCore::SearchFieldCancelButtonMac::SearchFieldCancelButtonMac):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.mm:
(WebCore::SearchFieldMac::SearchFieldMac):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.h:
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.mm:
(WebCore::SearchFieldResultsMac::SearchFieldResultsMac):
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
(WebCore::SliderThumbMac::SliderThumbMac):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::SliderTrackMac::SliderTrackMac):
(WebCore::SliderTrackMac::rectForBounds const):
(WebCore::SliderTrackMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextAreaMac.h:
* Source/WebCore/platform/graphics/mac/controls/TextAreaMac.mm:
(WebCore::TextAreaMac::TextAreaMac):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm:
(WebCore::TextFieldMac::TextFieldMac):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm:
(WebCore::ToggleButtonMac::ToggleButtonMac):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::updateControlPartForRenderer const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::encode): Deleted.
(IPC::ArgumentCoder<ControlPart>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/270445@main
